### DISCUSSION TITLE
YouTube API Getting Remaining Quota or Used Quota

### DISCUSSION THREAD

**Discussion by Firecul** [ThioJoe/YT-Spammer-Purge#688] [Q&A] [answered]
Related to #682 and many others, is it possible to poll the YT data API for the remaining quota?

If not could the app keep a rough track of API calls and warn when getting close to the limit?

**Comment by ThioJoe** [comment, marked as answer]
I've looked into this extensively, and the answer is "kind of", but it is incredibly convoluted. For some reason there is no "how much quota left" query, nor is there a "how much quota have I used" query. 

The only way it can be done is using another API, the google "cloud monitoring" API: 
- https://cloud.google.com/blog/topics/cost-management/how-to-programmatically-manage-quotas-in-google-cloud
- https://cloud.google.com/monitoring/alerts/using-quota-metrics

Though even with that, it basically requires you to set up a whole custom query using "MQL" which is an entire language just for report queries. I was sort of able to get it to create a chart that shows when you kinda go over the limit, but it's not even exact. I'm pretty sure it works by summing up every minute's worth of API usage over a 24 hour period. I haven't even really figured out how to do a text based query.

I essentially gave up on it, but I might look more into it again at some point.

Also as a reminder to myself and anyone else who stumbles upon this, "Service Usage" is _not_ what you're looking for. It only gives you the quota limit, not the amount you've actually used, as described by the blue 'starred' note here: https://cloud.google.com/service-usage/docs/manage-quota

> Note: Service Usage only displays information about a project's quota limits, not its quota usage. You can get quota usage data using the [Cloud Monitoring API](https://cloud.google.com/monitoring/docs/apis) and the [consumer_quota](https://cloud.google.com/monitoring/api/resources#tag_consumer_quota) resource.

The note then points to this which is part of the cloud monitoring API, but requires all the convoluted stuff I described above: https://cloud.google.com/monitoring/api/resources#tag_consumer_quota

Also I'm going to rename the title of this thread so it might turn on up google for some people, so they can save the hours it took me simply to find that blue starred note buried in the documentation.


This was the best I was able to do, which was make this chart showing usage and the limit:
<img width="851" alt="image" src="https://user-images.githubusercontent.com/12518330/154808491-e5fe225b-ea83-4f77-8cfe-fb3abdcb8a51.png">

**Notice though** how it seems to be more of a rolling average, which is not how the API actually resets quota. The quota resets all at once (at midnight PST I believe), so it should really only have instant dropoffs, and no "partial" dropoffs at any time. I think there might be ways to set a "static" time window instead of a rolling window, but I couldn't figure it out.

And this is the configuration for that chart:
<img width="331" alt="image" src="https://user-images.githubusercontent.com/12518330/154808587-e134c588-a6af-48c3-946f-d59a2064fd01.png"><img width="325" alt="image" src="https://user-images.githubusercontent.com/12518330/154808601-7066e602-79aa-4959-bf6c-1e2f3bed24c1.png"><img width="337" alt="image" src="https://user-images.githubusercontent.com/12518330/154808693-6461c32b-2fd3-4faa-9ca6-eee8ed542c74.png">

**Comment by Kurdakov** [comment]
I think it's possible to approximately compute used quota.
say, if previous time of quota is not saved, assume it's fresh quota, save time relative to PST time (if PST time reset of quota is correct) and get quota setting (I think that might be someone could manage to rise quota, so have your own quota setting might be useful), now if there is no crash
https://developers.google.com/youtube/v3/getting-started?hl=tr#calculating-quota-usage
each 100 fetched comments is 1 quota item.
each deleted/reported comment 50 quota item (if there is no verification, if verification of deletion +1 per deleted comment (?)), after running a round, save estimated used quota.

after script starts  anew - check saved time (and compare with current time shifted to PST time), if more than 1 day passed, think that quota is reset and save used quota  to zero.

based on feedback (marking feature as experimental) - assumptions might be corrected and estimation will be closer to that which youtube thinks is user quota.

**Reply by ThioJoe** [reply]
Yea I agree the best way to most accurately measure quota usage would probably be to have the program calculate it itself. 

There might be some exceptions / edge cases, like if the program crashes (in which case perhaps the program could just dump a log that can be read on the next run, if I really wanted to get fancy), or if a user is running multiple parallel instances of the program with the same google cloud project.

But I suppose it doesn't make sense to not make a feature just because of some edge cases that can be addressed later.
